### PR TITLE
Osembe

### PR DESCRIPTION
**Pull request by HauHe**
Function and test added to read ProductionByTechnologyAnnual data and create dataframe with primary energy by country.

### CODE REVIEW

File: resultify.py [hunk 1]
```diff
@@ -46,6 +46,29 @@ def filter_emission(df: pd.DataFrame, emission: List) -> pd.DataFrame:
...
+def filter_primary_energy(df: pd.DataFrame, technologies: List) -> pd.DataFrame:
+    """Return rows that indicate Primary Energy use/generation
+    """
+    for t in range(len(technologies)):
+        technologies[t] = technologies[t].replace('*','')
+    df['REGION'] = df['TECHNOLOGY'].str[:2]
+    df_f = pd.DataFrame(columns=['REGION','TECHNOLOGY','FUEL','YEAR','VALUE'])
+    for t in range(len(technologies)):
+        if len(technologies[t])==2:
+            mask = df['TECHNOLOGY'].str[2:4] == technologies[t]
```
Comment (willu47) [inline resultify.py:58]: Could you extract the position of the characters you match (`2:4`) from the `**` mask you pass in?  You wouldn't need the `if, else` logic then either.

File: test_resultify.py [hunk 2]
```diff
@@ -54,4 +54,32 @@ def test_filter_fuel(self):
...
 
+        pd.testing.assert_frame_equal(actual.set_index(index), expected.set_index(index), check_index_type=False)
+    
+class TestEnergy:
+
+    def test_filter_primary_energy(self):
+        filepath = os.path.join("tests","fixtures","ProductionByTechnologyAnnual.csv")
+        input_data = pd.read_csv(filepath)
+
+        technologies = ['****00I00','****00X00','**OC*****','**SO*****','**WI*****']
```
Comment (willu47) [inline test_resultify.py:65]: You might want to split this test up into multiple tests, one for each pattern which is matched?  What happens if there's an incorrect pattern passed in. For example `*********` or `**I*AM*TOO*LONG`?

Comment (willu47) [reply test_resultify.py:65]: Hint: to check that an error is raised in Pytest, you use the following:
```python
with raises(ValueError) as ex:
    filter_primary_energy(input_data, bogus_data)
assert str(ex) == "The error message which is raised"
```

**Review by willu47 — COMMENTED** [review]
Looks good @HauHe - I've made a few suggestions for tweaks on how to make this as general as we can.

File: resultify.py (hunk 1 above)
Comment (HauHe) [reply resultify.py:58]: Not sure, how to do so. The issue is that I filter to [2:4] when to account the energy from Solar, Wind and Ocean, but to letter [4:9] (could also be only letter 6) when accounting for oil, gas and coal.

File: test_resultify.py (hunk 2 above)
Comment (HauHe) [reply test_resultify.py:65]: Thanks for the tips! I will work on this.

File: resultify.py (hunk 1 above)
Comment (willu47) [reply resultify.py:58]: One idea is to use [Regular Expressions](https://en.wikipedia.org/wiki/Regular_expression) in the configuration file, and then [pandas.str.contains](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.contains.html).

You can try out reg ex patterns [here](https://regexr.com/)

Comment (HauHe) [reply resultify.py]: I gave it a go. Might be a bit more sophisticated to modify the config file. But the code is definitely simpler.

File: test_resultify.py (hunk 2 above)
Comment (HauHe) [reply test_resultify.py]: Well now the input are regex. Not sure how to test if a regex makes sense. Any suggestion?

**Comment by HauHe** [comment]
The previous commit closes issue #6

**Pull request merged**